sled-hardware/src/disk.rs

```diff
@@ -132,6 +132,46 @@ impl DiskPaths {
     }
 }
 
+// XXX MTZ: Make this an enum with DiskFirmware::Nvme?
+#[derive(
+    Debug, Clone, PartialEq, Eq, Hash, Ord, PartialOrd, Deserialize, Serialize,
+)]
+pub struct DiskFirmware {
+    active_slot: u8,
+    next_active_slot: Option<u8>,
+    slot1_read_only: bool,
```

Comment:
It feels odd that all the other fields imply an arbitrary number of slots, but this field specifically embeds the notion of "slot1"? Also, is this one or zero-indexed?

Reply:
Welcome to the NVMe spec! The spec states that devices have slot 1-7 (and potentially more in the future as there's unused padding at the end of the C struct). The spec also says that only slot 1 can be read only. Not every device will have all 7 slots available, for example the C bhyve nvme devices shows only one slot that is read only, while some of our WDC devices show 4 slots.  And not every available slot will necessarily contain firmware.  The firmware version itself is defined to be an ascii string hence the `Vec<Option<String>>`.

After having chatted with @jgallagher about this we landed on `NvmeSlot` being a wrapping type to address the array index being 0 but corresponding with slot1 - this can be seen [here] (https://github.com/oxidecomputer/libnvme/blob/mike/libnvme-firmware/libnvme/src/firmware.rs#L112-L135) and has yet to be officially reviewed.  The thought was that one could gain access to all the slots via an iterator and upstream of the library you could deal with the 0 indexed array. 

I originally had this represented as an enum but was unsure what the underlying CRDB type should look like.  Very open to suggestions here as I just went with _something_ for this PR as a placeholder.